components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/cache/AuthenticationContextLoader.java

```diff
@@ -0,0 +1,304 @@
+                FrameworkServiceDataHolder.getInstance().getApplicationManagementService();
+        try {
+            serviceProvider = applicationManager.getApplicationByResourceId(
+                    optApplicationConfig.getServiceProviderResourceId(), tenantDomain);
+            if (serviceProvider == null) {
+                throw new SessionDataStorageOptimizationClientException(
+                        String.format("Cannot find the Service Provider by the resource ID: %s tenant domain: %s",
+                                optApplicationConfig.getServiceProviderResourceId(), tenantDomain));
+            }
+            if (serviceProvider.getLocalAndOutBoundAuthenticationConfig() == null) {
+                throw new SessionDataStorageOptimizationClientException(
+                        String.format("Cannot find the LocalAndOutBoundAuthenticationConfig for the " +
+                                        "Service Provider by the resource ID: %s tenant domain: %s",
+                                optApplicationConfig.getServiceProviderResourceId(), tenantDomain));
+            }
+            serviceProvider.getLocalAndOutBoundAuthenticationConfig().setAuthenticationSteps(
+                    optApplicationConfig.getAuthenticationSteps(tenantDomain));
+        } catch (IdentityApplicationManagementException | FrameworkException e) {
+            throw new SessionDataStorageOptimizationServerException(
+                    String.format("Error occurred while retrieving the service provider by resource id: %s " +
+                            "tenant domain: %s", optApplicationConfig.getServiceProviderResourceId(), tenantDomain), e);
+        }
+        return serviceProvider;
+    }
+
+    private IdentityProvider getIdPByIdPName(String idPName, String tenantDomain)
+            throws SessionDataStorageOptimizationException {
+
+        IdentityProviderManager manager =
+                (IdentityProviderManager) FrameworkServiceDataHolder.getInstance().getIdentityProviderManager();
+        IdentityProvider idp;
+        try {
+            idp = manager.getIdPByName(idPName, tenantDomain);
+            if (idp == null) {
+                throw new SessionDataStorageOptimizationClientException(String.format(
+                        "Cannot find the Identity Provider by the name: %s tenant domain: %s", idPName, tenantDomain));
+            }
+        } catch (IdentityProviderManagementException e) {
+            throw new SessionDataStorageOptimizationServerException(String.format(
+                    "Failed to get the Identity Provider by name: %s tenant domain: %s", idPName, tenantDomain), e);
+        }
+        return idp;
+    }
+
+    private IdentityProvider getIdPByResourceID(String resourceId, String tenantDomain)
+            throws SessionDataStorageOptimizationException {
+
+        if (StringUtils.isEmpty(resourceId)) {
+            throw new SessionDataStorageOptimizationClientException("Resource Id is null. " +
```

Comment:
Whats the reason for throwing an exception here as oppose to simply returning null?

Reply:
During optimization, we are not storing the entire IDP object. We only store the necessary attributes to load it back to the cache immediately. So, the DB should ideally have the correct ID. When we are retrieving from DB, in order to load the correct entire object back we need the resource ID. Returning null will result in not retrieving the full session context cache from the optimized session context cache.

Reply:
Having it as an exception seems fine in that case. But can we say that this is a client exception? everything used here as input is also from the server itself right?

Reply:
Ideally it should not be null. However, yes we should not exactly refer it as a client exception.
It was noticed that that this is actually a redundant check as we are checking [earlier](https://github.com/wso2/carbon-identity-framework/blob/e784a63a2be2c959d9af0701cf43399d07cb90fd/components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/cache/AuthenticationContextLoader.java#L147) if the resource id is null as well. So will remove this.

Reply:
Resolved with [`4b4997e`](https://github.com/wso2/carbon-identity-framework/pull/4446/commits/4b4997ee7009696b196bf176bbc1702e07450522)